Replace `ConsumedNoteInfo` with `Nullifier`

bobbinth opened the pull request:
This PR addresses a part of #350. Specifically, I removed `ConsumedNoteInfo` struct and replaced it with `Nullifier`. Using only nullifiers to compute consumed note commitment is sufficient since #283 as the verifier no longer needs to know script roots for the consumed notes.

The commitment to consumed notes is still computed as a sequential hash of `(nullifier, ZERO)` tuples instead of just as a sequential hash of nullifiers. I kept it this way primarily to keep the scope of changes relatively small and also because we may need to update the logic again to support #353.

bobbinth commented:
cc @igamigo: this might affect some things in the client, but the changes should be very minor.

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -656,7 +656,7 @@ end
 #! Process the consumed notes data provided via the advice provider. This involves reading the data
```
Somehow it is weird to say `Process the consumed notes` before the notes are being consumed. I would have expected to call those notes `input notes` or `notes to be consumed`

bobbinth replied:
Yeah - `input notes` would have been better. We'll move towards this terminology over time.

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
 proc.process_consumed_notes_data
     # load the consumed notes data onto the advice stack
     exec.layout::get_nullifier_com adv.push_mapval dropw
+    # => [...]
```
Nice to add those here, that helps and is consistent

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
 
-    # assert the number of consumed notes is within limits
-    dup exec.constants::get_max_num_consumed_notes lte assert
+    # assert the number of consumed notes is within limits; since max number of consumed notes is
```
`since max number of consumed notes is smaller than 2^32` 

Isn't that what you check here? If so, shouldn't you say then:

`check that max number of consumed notes is smaller than 2^32`

bobbinth replied:
Because we expected the number of notes to be smaller than $2^{32}$ we can use a combination of `u32assert2` and `u32lte` instead of a general `lte` (which would have been less efficient).

I re-worded the statement to make this more clear.

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
-    dup exec.constants::get_max_num_consumed_notes lte assert
+    # assert the number of consumed notes is within limits; since max number of consumed notes is
+    # smaller than 2^32, we can use a more efficient u32 comparison
+    dup exec.constants::get_max_num_consumed_notes u32assert2 u32lte assert
```
why does `u32assert2` only need 1 cycle whereas `u32assert` needs 3?

bobbinth replied:
`u32assert2` is a native VM operation while `u32assert` is emulated as `push.0 u32assert2 drop`. The main reason for making `u32assert2` operation native was that we frequently need to assert that top 2 values on the stack are `u32` values (more frequently than we need to assert that top value is `u32`).

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
 
     # initialize counter for consumed notes
     push.0
+    # => [num_processed_notes, num_notes, ...]
```
maybe write `=> [num_processed_notes = 0, num_notes, ...]`

bobbinth replied:
Agreed. Changed!

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
     # loop over consumed notes and read data
     # ---------------------------------------------------------------------------------------------
 
     # initialize counter for consumed notes
```
here you say the counter is for consumed notes, the name of the counter is processed notes

bobbinth replied:
Yep. Fixed this.

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
     push.0
+    # => [num_processed_notes, num_notes, ...]
 
     # check if the number of consumed notes is greater then 0. Conditional for the while loop.
```
Is the `number of consumed notes` the `num_notes` ? `consumed` somehow implied that those notes were already processed. To me, notes to be consumed makes more sense.

bobbinth replied:
Yeah - I think the input/output note terminology would have been better. We should move in that direction, but it'll take some time before we get everything renamed.

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
 
         # increment consumed note counter and check if we should loop again
         add.1 dup.1 dup.1 neq
+        # => [has_more_notes, num_processed_notes + 1, num_notes, ...]
```
in the comment you call the counter consume note counter in the the comment below you call it `num_processed_notes`

bobbinth replied:
Good point! Changed.

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
```diff
@@ -679,78 +679,99 @@ end
 
     # initiate counter for nullifiers
     push.0
+    # => [num_processed_notes, num_notes, ...]
```
can you call this counter `num_nullifiers` or so?

bobbinth replied:
We are still technically iterating over notes (and just extracting a nullifier from a note at each iteration) - so, I'd keep it as is.

Dominik1999 commented on miden-lib/asm/miden/sat/internal/prologue.masm:
There is no real MASM code change, right? The only change is an optimization using `u32assert2`. This is because the MASM code hashes sequentially whatever is there, and now we hash `ZERO` instead of a `script_root`.

bobbinth replied:
The main change is to replace `script_root` with `ZERO` - without this change the test would be failing.

Dominik1999 commented on miden-tx/src/verifier/mod.rs:
```diff
@@ -4,8 +4,8 @@ use miden_lib::outputs::{
 use miden_objects::{
-    notes::NoteEnvelope,
-    transaction::{ConsumedNoteInfo, ProvenTransaction},
+    notes::{NoteEnvelope, Nullifier},
```
this renaming makes a lot of sense to me

Dominik1999 commented on miden-tx/src/verifier/mod.rs:
```diff
@@ -58,12 +58,13 @@ impl TransactionVerifier {
+
     /// Returns the consumed notes commitment.
-    fn compute_consumed_notes_hash(consumed_notes: &[ConsumedNoteInfo]) -> Digest {
+    fn compute_consumed_notes_hash(consumed_notes: &[Nullifier]) -> Digest {
```
Are those again notes to be consumed in a transaction or notes that were already consumed, i.e., the transaction is over?

bobbinth replied:
I think these are the same thing - the set of notes to be consumed in a transaction is the same as the set of notes consumed when a transaction is over.

Dominik1999 commented on miden-tx/src/verifier/mod.rs:
```diff
@@ -58,12 +58,13 @@ impl TransactionVerifier {
+
     /// Returns the consumed notes commitment.
-    fn compute_consumed_notes_hash(consumed_notes: &[ConsumedNoteInfo]) -> Digest {
+    fn compute_consumed_notes_hash(consumed_notes: &[Nullifier]) -> Digest {
```
Should you rename this variable `consumed_notes` if it only contains an array of `Nullifiers`?

bobbinth replied:
Possibly - but I'll tackle this in a different PR when I refactor the structure to use input/output note terminology.

Dominik1999 commented on miden-tx/src/verifier/mod.rs:
```diff
@@ -58,12 +58,13 @@ impl TransactionVerifier {
+
     /// Returns the consumed notes commitment.
-    fn compute_consumed_notes_hash(consumed_notes: &[ConsumedNoteInfo]) -> Digest {
+    fn compute_consumed_notes_hash(consumed_notes: &[Nullifier]) -> Digest {
```
Should you rename this variable `consumed_notes` if it only contains an array of `Nullifiers`?

bobbinth replied:
Possibly - but I'll tackle this in a different PR when I refactor the structure to use input/output note terminology.

Dominik1999 commented on objects/src/notes/nullifier.rs:
```diff
@@ -0,0 +1,105 @@
+///
+/// This achieves the following properties:
+/// - Every note can be reduced to a single unique nullifier.
+/// - We cannot derive a note's hash from its nullifier.
```
`We cannot derive a note's hash from its nullifier.` 

Shouldn't be the other way around important? That we cannot derive the note's nullifier from its hash? The hash is public knowledge, but the nullifier should be hidden to ensure private consumption, right?

Dominik1999 replied:
or both is important

bobbinth replied:
Both are important - I'll change this.

Dominik1999 commented on objects/src/transaction/proven_tx.rs:
```diff
@@ -20,7 +19,7 @@ pub struct ProvenTransaction {
     initial_account_hash: Digest,
     final_account_hash: Digest,
-    consumed_notes: Vec<ConsumedNoteInfo>,
+    consumed_notes: Vec<Nullifier>,
```
I understand that `consumed_notes` are defined by their nullifiers. But why don't we just name the variable as is, as `nullifiers`

bobbinth replied:
I think nullifiers would not be descriptive enough. We could use `consumed_note_nullifiers` but that's a but too long.

Dominik1999 reviewed (approved):
Looks good to me. Left some comments inline. 

And why do you not need to change a test? Isn't there a test for consuming notes?

bobbinth commented:
> And why do you not need to change a test? Isn't there a test for consuming notes?

Because I've changed both the Rust code and the MASM code to compute the input note has in the same way, the tests don't need to be changed (if I updated only MASM or only Rust, the tests would be failing).
